nomad/fsm.go

```diff
@@ -334,14 +341,16 @@ func (n *nomadFSM) applyUpsertNode(buf []byte, index uint64) interface{} {
 	return nil
 }
 
-func (n *nomadFSM) applyDeregisterNode(buf []byte, index uint64) interface{} {
+func (n *nomadFSM) applyDeregisterNode(reqType structs.MessageType, buf []byte, index uint64) interface{} {
 	defer metrics.MeasureSince([]string{"nomad", "fsm", "deregister_node"}, time.Now())
 	var req structs.NodeDeregisterRequest
 	if err := structs.Decode(buf, &req); err != nil {
 		panic(fmt.Errorf("failed to decode request: %v", err))
 	}
 
-	if err := n.state.DeleteNode(index, []string{req.NodeID}); err != nil {
+	ctx := context.WithValue(context.Background(), state.CtxMsgType, reqType)
+
+	if err := n.state.DeleteNodeCtx(ctx, index, []string{req.NodeID}); err != nil {
```

Comment:
Why use a context instead of passing the type as an argument explicitly?

A context seems easier to get wrong and the FSM is our most critical code path. I think it will also cause at least a couple extra heap allocations which isn't a blocker, but this is one of our most performance sensitive areas as well.

Reply:
yeah I don't think context turned out as nice as I thought it would initially, backed out in favor of passing in message type